File: src/python/serial_port_finder.py
import sys
import glob
import serial.tools.list_ports

def find_usb_port():
    if sys.platform.startswith('win'):
        # Windows: use pyserial's list_ports to find COM ports
        ports = list(serial.tools.list_ports.comports())
        for port in ports:
            if 'USB' in port.description:
                return port.device
        raise RuntimeError("No USB COM port found on Windows")
    else:
        # Unix/macOS
        ports = glob.glob('/dev/tty.usbmodem*')
        if ports:
            return ports[0]
        raise RuntimeError("No /dev/tty.usbmodem* device found on Unix/macOS")

File: src/R/serial_port_finder.R
find_usb_port <- function() {
  sysname <- Sys.info()[["sysname"]]

  if (sysname == "Windows") {
    ports <- try(serial::listPorts(), silent = TRUE)
    if (inherits(ports, "try-error") || length(ports) == 0) {
      stop("No serial ports found on Windows")
    }
    usb_ports <- ports[grepl("usb", ports, ignore.case = TRUE)]
    if (length(usb_ports) == 0) {
      stop("No USB serial port found on Windows")
    }
    cat("Found USB port on Windows:", usb_ports[1], "\n")
    return(usb_ports[1])

  } else if (sysname %in% c("Darwin", "Linux")) {
    usb_ports <- Sys.glob("/dev/tty.usbmodem*")
    if (length(usb_ports) == 0) {
      usb_ports <- Sys.glob("/dev/ttyUSB*")
    }
    if (length(usb_ports) == 0) {
      stop("No /dev/tty.usbmodem* or /dev/ttyUSB* device found on Unix/macOS/Linux")
    }
    cat("Found USB port on Unix/macOS/Linux:", usb_ports[1], "\n")
    return(usb_ports[1])

  } else {
    stop("Unsupported OS for USB port detection")
  }
}
